quickwit-search/src/export/mod.rs

```diff
@@ -0,0 +1,123 @@
+// Quickwit
+//  Copyright (C) 2021 Quickwit Inc.
+//
+//  Quickwit is offered under the AGPL v3.0 and as commercial software.
+//  For commercial licensing, contact us at hello@quickwit.io.
+//
+//  AGPL:
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+mod collector;
+mod leaf;
+mod root;
+
+pub use collector::{FastFieldCollector, FastFieldCollectorBuilder};
+pub use leaf::leaf_export;
+pub use root::root_export;
+
+use serde::Deserialize;
+use std::{fmt::Display, io::Write};
+use tracing::info;
+
+use crate::SearchError;
+
+/// Output format for export.
+#[derive(Deserialize, Debug, PartialEq, Eq, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+pub enum OutputFormat {
+    /// Format data by row in binary format. TODO: add link to format.
+    RowBinary,
+    /// Comma Separated Values format (https://datatracker.ietf.org/doc/html/rfc4180).
+    /// By default, the delimiter is ,.
+    CSV,
+}
+
+impl Default for OutputFormat {
+    fn default() -> Self {
+        OutputFormat::RowBinary
+    }
+}
+
+impl ToString for OutputFormat {
+    fn to_string(&self) -> String {
+        match &self {
+            Self::RowBinary => "rowbinary".to_string(),
+            Self::CSV => "csv".to_string(),
+        }
+    }
+}
+
+impl From<String> for OutputFormat {
+    fn from(spec: String) -> Self {
+        match spec.as_str() {
+            "rowbinary" => Self::RowBinary,
+            _ => Self::CSV,
+        }
+    }
+}
+
+pub struct ExportSerializer<'a> {
+    output_format: OutputFormat,
+    writer: &'a mut Vec<u8>,
+}
+
+impl<'a> ExportSerializer<'a> {
+    pub fn write_u64(&mut self, values: Vec<u64>) -> crate::Result<()> {
```

Comment:
you can remove the distinction u64/i64 by cheating a bit since FastValue has a `.as_u64()` method.

Reply:
I have several issues here:

This function seems lowlevel enough that it should not have any concern about SearchError.
If it had to return an error, it should probably be something like io::Result<()>. The high level calling code should then convert the error to a higher level error if needed.

BUT:
your code cannot error in reality. We are writing in a Vec anyway.

The trouble is coming from the fact that you are using io::Write.
A better way (the compiler might even spew better code although I doubt it) is to use extend.

Finally there is technically a bug in your code.
`.write()` can actually only write you content partially. You almost always want to call `.write_all()`.

So after rewrite

```rust
#[inline(never)]
fn serialize_row_binary<TFastValue: FastValue>(vals: &[TFastValue], buffer: &mut Vec<u8>) {
    buffer.clear(); // by the way we need to comment in the public function that it clears the buffer if it is not empty
    buffer.reserve_exact(std::mem::size_of::<TFastValue>() * vals.len());
    for &val in vals {
        buffer.extend(val.as_u64().to_le_bytes());
    }
}
```

Reply:
(I tested in godbolt, rust is unfortunately not smart enough to transform that in a memcopy. If the serialization is not negligeable, we can special case for little endian arch.